Gan training

MylesBartlett opened the pull request.

Commit:
Added GAN training routine.

Commit:
Added function to apply spectral_norm to all conv/fc layers recursively.
Added GAN discriminator.

Commit:
Formatting.

Commit:
Added discriminator to init.

Commit:
Added GAN training script.

Commit:
Added GAN training script.

Commit:
Fixed disc_optimizer constructor.

Commit:
Fixed in_channels attribute

Commit:
Fixed in_channels attribute

Commit:
Updated initialization.

Commit:
Corrected loss

Commit:
generator pretraining

Commit:
Added pretraining command line argument.

Commit:
Added spectral normalization back.

Commit:
Added summary writer condition.

Commit:
Formatting.

Commit:
Corrected Complex Model

Commit:
Made discriminator blocks private.

Commit:
Default argument change.

Commit:
Default argument change.

Commit:
Debugging.

Commit:
Init

Commit:
Name change

Commit:
Debugging

Commit:
Name change.

Commit:
Undo debugging changes.

Commit:
Pre-training print statement

Commit:
Changed default discriminator model to GatedDiscriminator.
Made it so the individual generator losses are logged in addition to the total loss.
Added command line argument for weight on adversarial loss.

Commit:
Removed unused imports.

Commit:
Changed disc strides.

Commit:
Removed print statement.

Commit:
Changed default discriminator.

Commit:
Fixed simple discriminator.

Commit:
debugging

Commit:
debugging

Commit:
Wrong argument to spectral norm

Commit:
Added adv_weight argument

Commit:
Fixed disc argument.

Commit:
Uncommented global pooling

Commit:
int cast

Commit:
Removed redundant code.

Commit:
Made it so the generator is called only once each epoch.

Commit:
Comments

Commit:
typo

Commit:
Eval fix

Commit:
Eval fix

Commit:
Eval fix

Commit:
Show weighted adversarial loss.

Commit:
Updated default parameters.

Commit:
Changed optimizer to Adam (always). Added argument for beta values.

Commit:
Updated GatedDiscriminator.

tmke8 commented on main_gan.py:
```diff
@@ -0,0 +1,256 @@
+    # CNN
+    parser.add_argument('--cnn_in_channels', type=int, default=3)
+    parser.add_argument('--cnn_hidden_channels', type=int, default=32)
+    parser.add_argument('--cnn_num_hidden_layers', type=int, default=7)
```
I think you have to change it here now to `cnn_hidden_layers`

MylesBartlett replied:
Just did that, anticipating your comment :)

MylesBartlett committed:
Update main_gan.py

Changed cnn_num_hidden_layers to cnn_hidden_layers, in accordance with the pending pull request.

tmke8 commented on models/simple_complex_cnn.py:
```diff
@@ -19,7 +19,7 @@ def __init__(self, args):
         # Hidden layers
         for _ in range(args.cnn_num_hidden_layers):
-            layers.append(ComplexGatedConvLayer(args.cnn_hidden_channels, args.cnn_hidden_channels,
+            layers.append(ComplexGatedConvLayer(num_hidden_channels, num_hidden_channels,
```
This change is causing a conflict because this line was also changed on master.

MylesBartlett replied:
The hidden layers should take the local variable 'num_hidden_channels' as the argument to their constructor. I only have access to my phone atm with which it's not easy to resolve conflicts due to the formatting.

tmke8 replied:
Ok, I can resolve the conflict

MylesBartlett replied:
Thanks

tmke8 replied:
done

tmke8 committed:
Merge branch 'master' into gan_training

MylesBartlett committed:
Merge branch 'master' into gan_training

tmke8 commented on main_gan.py:
```diff
@@ -0,0 +1,256 @@
+import models
+
+
+def parse_arguments(raw_args=None):
```
All the code duplication is not nice (makes it very hard to maintain). Can you at least import this function from main.py? I think it should be the same there.

MylesBartlett replied:
The arguments for main_gan.py are different.
